File: scripts/chess/chess-helper.js
#!/usr/bin/env node
/**
 * Chess Helper — CLI tool for Synapse's chess domain adapter.
 * Opens positions on Lichess, validates FEN, renders boards.
 *
 * Usage:
 *   node chess-helper.js validate --fen "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
 *   node chess-helper.js url --fen "<FEN>" [--color black]
 *   node chess-helper.js open --fen "<FEN>" [--color black]
 *   node chess-helper.js board --fen "<FEN>" [--color black]
 *   node chess-helper.js position --white "Ke1,Qd1" --black "Ke8" [--to-move white] [--castling KQkq]
 */

const { execSync } = require("child_process");

function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i].startsWith("--")) {
      const key = argv[i].slice(2);
      const val = argv[i + 1] && !argv[i + 1].startsWith("--") ? argv[i + 1] : true;
      args[key] = val;
      if (val !== true) i++;
    } else {
      args._.push(argv[i]);
    }
  }
  return args;
}

// --- FEN validation (structural only, v1) ---

const PIECE_CHARS = "pnbrqkPNBRQK";
const WHITE_PIECES = { K: "King", Q: "Queen", R: "Rook", B: "Bishop", N: "Knight", P: "Pawn" };
const BLACK_PIECES = { k: "King", q: "Queen", r: "Rook", b: "Bishop", n: "Knight", p: "Pawn" };

function validateFen(fen) {
  const fields = fen.trim().split(/\s+/);
  if (fields.length < 1 || fields.length > 6) {
    return { valid: false, error: `FEN must have 1-6 fields, got ${fields.length}` };
  }

  // Fill defaults for missing fields
  const placement = fields[0];
  const sideToMove = fields[1] || "w";
  const castling = fields[2] || "-";
  const enPassant = fields[3] || "-";
  const halfmove = fields[4] || "0";
  const fullmove = fields[5] || "1";

  // Validate piece placement
  const ranks = placement.split("/");
  if (ranks.length !== 8) {
    return { valid: false, error: `Expected 8 ranks, got ${ranks.length}` };
  }

  let whitePieces = 0;
  let blackPieces = 0;

  for (let i = 0; i < 8; i++) {
    const rank = ranks[i];
    let squares = 0;
    for (const ch of rank) {
      if (ch >= "1" && ch <= "8") {
        squares += parseInt(ch, 10);
      } else if (PIECE_CHARS.includes(ch)) {
        squares += 1;
        if (ch === ch.toUpperCase()) whitePieces++;
        else blackPieces++;
      } else {
        return { valid: false, error: `Invalid character '${ch}' in rank ${i + 1}` };
      }
    }
    if (squares !== 8) {
      return { valid: false, error: `Rank ${i + 1} has ${squares} squares (expected 8)` };
    }
  }

  // Validate side to move
  if (sideToMove !== "w" && sideToMove !== "b") {
    return { valid: false, error: `Side to move must be 'w' or 'b', got '${sideToMove}'` };
  }

  // Validate castling
  if (castling !== "-" && !/^[KQkq]{1,4}$/.test(castling)) {
    return { valid: false, error: `Invalid castling rights: '${castling}'` };
  }

  // Validate en passant
  if (enPassant !== "-" && !/^[a-h][36]$/.test(enPassant)) {
    return { valid: false, error: `Invalid en passant square: '${enPassant}'` };
  }

  // Validate halfmove and fullmove clocks
  if (!/^\d+$/.test(halfmove) || parseInt(halfmove, 10) < 0) {
    return { valid: false, error: `Invalid halfmove clock: '${halfmove}'` };
  }
  if (!/^\d+$/.test(fullmove) || parseInt(fullmove, 10) < 1) {
    return { valid: false, error: `Invalid fullmove number: '${fullmove}'` };
  }

  const fullFen = `${placement} ${sideToMove} ${castling} ${enPassant} ${halfmove} ${fullmove}`;

  return {
    valid: true,
    fen: fullFen,
    pieces: { white: whitePieces, black: blackPieces },
    side_to_move: sideToMove === "w" ? "white" : "black",
    castling,
    en_passant: enPassant,
  };
}

// --- URL generation ---

function buildUrl(fen, color) {
  const urlFen = fen.replace(/ /g, "_");
  let url = `https://lichess.org/analysis/standard/${urlFen}`;
  if (color === "black") url += "?color=black";
  return url;
}

// --- Board rendering ---

const UNICODE_PIECES = {
  K: "\u2654", Q: "\u2655", R: "\u2656", B: "\u2657", N: "\u2658", P: "\u2659",
  k: "\u265A", q: "\u265B", r: "\u265C", b: "\u265D", n: "\u265E", p: "\u265F",
};

function renderBoard(fen, color) {
  const placement = fen.split(/\s+/)[0];
  const ranks = placement.split("/");
  const board = [];

  for (const rank of ranks) {
    const row = [];
    for (const ch of rank) {
      if (ch >= "1" && ch <= "8") {
        for (let j = 0; j < parseInt(ch, 10); j++) row.push("\u00B7");
      } else {
        row.push(UNICODE_PIECES[ch] || ch);
      }
    }
    board.push(row);
  }

  const lines = [];
  if (color === "black") {
    for (let r = 7; r >= 0; r--) {
      const rankNum = r + 1;
      lines.push(`  ${rankNum}  ${board[8 - 1 - r].slice().reverse().join(" ")}`);
    }
    lines.push("     h g f e d c b a");
  } else {
    for (let r = 0; r < 8; r++) {
      const rankNum = 8 - r;
      lines.push(`  ${rankNum}  ${board[r].join(" ")}`);
    }
    lines.push("     a b c d e f g h");
  }

  return lines.join("\n");
}

// --- Position from piece lists ---

const FILES = "abcdefgh";

function parsePieceList(list) {
  const pieces = [];
  for (const entry of list.split(",")) {
    const trimmed = entry.trim();
    if (!trimmed) continue;

    // Format: Xe1 where X is piece letter (K,Q,R,B,N,P) and e1 is square
    if (trimmed.length < 3 || trimmed.length > 3) {
      throw new Error(`Invalid piece entry '${trimmed}': expected format like 'Ke1'`);
    }

    const piece = trimmed[0];
    if (!"KQRBNPkqrbnp".includes(piece.toUpperCase())) {
      throw new Error(`Invalid piece letter '${piece}' in '${trimmed}'`);
    }

    const file = trimmed[1];
    const rank = trimmed[2];
    if (!FILES.includes(file) || rank < "1" || rank > "8") {
      throw new Error(`Invalid square '${file}${rank}' in '${trimmed}'`);
    }

    pieces.push({ piece, file: FILES.indexOf(file), rank: parseInt(rank, 10) - 1 });
  }
  return pieces;
}

function positionToFen(whitePieces, blackPieces, toMove, castling) {
  // Initialize empty board (8x8, rank 0 = rank 8 in chess)
  const board = Array.from({ length: 8 }, () => Array(8).fill(null));

  const occupied = new Set();

  function place(pieces, isWhite) {
    for (const p of pieces) {
      const boardRank = 7 - p.rank; // rank 1 = index 7, rank 8 = index 0
      const key = `${p.file},${boardRank}`;
      if (occupied.has(key)) {
        throw new Error(`Duplicate square: ${FILES[p.file]}${p.rank + 1}`);
      }
      occupied.add(key);
      board[boardRank][p.file] = isWhite ? p.piece.toUpperCase() : p.piece.toLowerCase();
    }
  }

  place(whitePieces, true);
  place(blackPieces, false);

  // Serialize to FEN
  const fenRanks = [];
  for (let r = 0; r < 8; r++) {
    let rankStr = "";
    let empty = 0;
    for (let f = 0; f < 8; f++) {
      if (board[r][f]) {
        if (empty > 0) { rankStr += empty; empty = 0; }
        rankStr += board[r][f];
      } else {
        empty++;
      }
    }
    if (empty > 0) rankStr += empty;
    fenRanks.push(rankStr);
  }

  const side = toMove === "black" ? "b" : "w";
  const castlingStr = castling || "-";
  return `${fenRanks.join("/")} ${side} ${castlingStr} - 0 1`;
}

// --- Commands ---

function cmdValidate(args) {
  if (!args.fen) {
    console.error(JSON.stringify({ error: "Missing --fen argument" }));
    process.exit(1);
  }
  const result = validateFen(args.fen);
  console.log(JSON.stringify(result, null, 2));
}

function cmdUrl(args) {
  if (!args.fen) {
    console.error(JSON.stringify({ error: "Missing --fen argument" }));
    process.exit(1);
  }
  const validation = validateFen(args.fen);
  if (!validation.valid) {
    console.error(JSON.stringify({ error: validation.error }));
    process.exit(1);
  }
  const url = buildUrl(validation.fen, args.color);
  console.log(JSON.stringify({ url, fen: validation.fen }, null, 2));
}

function cmdOpen(args) {
  if (!args.fen) {
    console.error(JSON.stringify({ error: "Missing --fen argument" }));
    process.exit(1);
  }
  const validation = validateFen(args.fen);
  if (!validation.valid) {
    console.error(JSON.stringify({ error: validation.error }));
    process.exit(1);
  }
  const url = buildUrl(validation.fen, args.color);

  let openCmd;
  switch (process.platform) {
    case "darwin": openCmd = "open"; break;
    case "win32": openCmd = "start"; break;
    default: openCmd = "xdg-open"; break;
  }

  try {
    execSync(`${openCmd} "${url}"`, { stdio: "ignore" });
    console.log(JSON.stringify({ opened: true, url, platform: process.platform }, null, 2));
  } catch (err) {
    console.log(JSON.stringify({
      opened: false,
      url,
      platform: process.platform,
      error: `Failed to open browser. Visit: ${url}`,
    }, null, 2));
  }
}

function cmdBoard(args) {
  if (!args.fen) {
    console.error(JSON.stringify({ error: "Missing --fen argument" }));
    process.exit(1);
  }
  const validation = validateFen(args.fen);
  if (!validation.valid) {
    console.error("Error: " + validation.error);
    process.exit(1);
  }
  console.log(renderBoard(validation.fen, args.color));
}

function cmdPosition(args) {
  if (!args.white && !args.black) {
    console.error(JSON.stringify({ error: "Provide at least --white or --black piece list" }));
    process.exit(1);
  }

  const whitePieces = args.white ? parsePieceList(args.white) : [];
  const blackPieces = args.black ? parsePieceList(args.black) : [];
  const toMove = args["to-move"] || "white";
  const castling = args.castling || undefined;

  const fen = positionToFen(whitePieces, blackPieces, toMove, castling);
  const url = buildUrl(fen, undefined);

  console.log(JSON.stringify({
    fen,
    url,
    pieces: { white: whitePieces.length, black: blackPieces.length },
  }, null, 2));
}

// --- Main ---

const args = parseArgs(process.argv);
const command = args._[0];

if (!command) {
  console.error("Usage: chess-helper.js <validate|url|open|board|position> [options]");
  process.exit(1);
}

try {
  switch (command) {
    case "validate": cmdValidate(args); break;
    case "url": cmdUrl(args); break;
    case "open": cmdOpen(args); break;
    case "board": cmdBoard(args); break;
    case "position": cmdPosition(args); break;
    default:
      console.error(`Unknown command: ${command}`);
      process.exit(1);
  }
} catch (err) {
  console.error(JSON.stringify({ error: err.message }));
  process.exit(1);
}

File: scripts/chess/chess-helper.py
#!/usr/bin/env python3
"""
Chess Helper — CLI tool for Synapse's chess domain adapter.
Opens positions on Lichess, validates FEN, renders boards.

Usage:
  python3 chess-helper.py validate --fen "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
  python3 chess-helper.py url --fen "<FEN>" [--color black]
  python3 chess-helper.py open --fen "<FEN>" [--color black]
  python3 chess-helper.py board --fen "<FEN>" [--color black]
  python3 chess-helper.py position --white "Ke1,Qd1" --black "Ke8" [--to-move white] [--castling KQkq]
"""

import argparse
import json
import platform
import subprocess
import sys

PIECE_CHARS = "pnbrqkPNBRQK"
FILES = "abcdefgh"

UNICODE_PIECES = {
    "K": "\u2654", "Q": "\u2655", "R": "\u2656", "B": "\u2657", "N": "\u2658", "P": "\u2659",
    "k": "\u265a", "q": "\u265b", "r": "\u265c", "b": "\u265d", "n": "\u265e", "p": "\u265f",
}


# --- FEN validation (structural only, v1) ---

def validate_fen(fen):
    fields = fen.strip().split()
    if len(fields) < 1 or len(fields) > 6:
        return {"valid": False, "error": f"FEN must have 1-6 fields, got {len(fields)}"}

    placement = fields[0]
    side_to_move = fields[1] if len(fields) > 1 else "w"
    castling = fields[2] if len(fields) > 2 else "-"
    en_passant = fields[3] if len(fields) > 3 else "-"
    halfmove = fields[4] if len(fields) > 4 else "0"
    fullmove = fields[5] if len(fields) > 5 else "1"

    ranks = placement.split("/")
    if len(ranks) != 8:
        return {"valid": False, "error": f"Expected 8 ranks, got {len(ranks)}"}

    white_pieces = 0
    black_pieces = 0

    for i, rank in enumerate(ranks):
        squares = 0
        for ch in rank:
            if "1" <= ch <= "8":
                squares += int(ch)
            elif ch in PIECE_CHARS:
                squares += 1
                if ch.isupper():
                    white_pieces += 1
                else:
                    black_pieces += 1
            else:
                return {"valid": False, "error": f"Invalid character '{ch}' in rank {i + 1}"}
        if squares != 8:
            return {"valid": False, "error": f"Rank {i + 1} has {squares} squares (expected 8)"}

    if side_to_move not in ("w", "b"):
        return {"valid": False, "error": f"Side to move must be 'w' or 'b', got '{side_to_move}'"}

    import re
    if castling != "-" and not re.match(r"^[KQkq]{1,4}$", castling):
        return {"valid": False, "error": f"Invalid castling rights: '{castling}'"}

    if en_passant != "-" and not re.match(r"^[a-h][36]$", en_passant):
        return {"valid": False, "error": f"Invalid en passant square: '{en_passant}'"}

    if not halfmove.isdigit() or int(halfmove) < 0:
        return {"valid": False, "error": f"Invalid halfmove clock: '{halfmove}'"}

    if not fullmove.isdigit() or int(fullmove) < 1:
        return {"valid": False, "error": f"Invalid fullmove number: '{fullmove}'"}

    full_fen = f"{placement} {side_to_move} {castling} {en_passant} {halfmove} {fullmove}"

    return {
        "valid": True,
        "fen": full_fen,
        "pieces": {"white": white_pieces, "black": black_pieces},
        "side_to_move": "white" if side_to_move == "w" else "black",
        "castling": castling,
        "en_passant": en_passant,
    }


# --- URL generation ---

def build_url(fen, color=None):
    url_fen = fen.replace(" ", "_")
    url = f"https://lichess.org/analysis/standard/{url_fen}"
    if color == "black":
        url += "?color=black"
    return url


# --- Board rendering ---

def render_board(fen, color=None):
    placement = fen.split()[0]
    ranks = placement.split("/")
    board = []

    for rank in ranks:
        row = []
        for ch in rank:
            if "1" <= ch <= "8":
                row.extend(["\u00b7"] * int(ch))
            else:
                row.append(UNICODE_PIECES.get(ch, ch))
        board.append(row)

    lines = []
    if color == "black":
        for r in range(7, -1, -1):
            rank_num = r + 1
            lines.append(f"  {rank_num}  {' '.join(reversed(board[7 - r]))}")
        lines.append("     h g f e d c b a")
    else:
        for r in range(8):
            rank_num = 8 - r
            lines.append(f"  {rank_num}  {' '.join(board[r])}")
        lines.append("     a b c d e f g h")

    return "\n".join(lines)


# --- Position from piece lists ---

def parse_piece_list(piece_list):
    pieces = []
    for entry in piece_list.split(","):
        trimmed = entry.strip()
        if not trimmed:
            continue

        if len(trimmed) != 3:
            raise ValueError(f"Invalid piece entry '{trimmed}': expected format like 'Ke1'")

        piece = trimmed[0]
        if piece.upper() not in "KQRBNP":
            raise ValueError(f"Invalid piece letter '{piece}' in '{trimmed}'")

        file_ch = trimmed[1]
        rank_ch = trimmed[2]
        if file_ch not in FILES or rank_ch < "1" or rank_ch > "8":
            raise ValueError(f"Invalid square '{file_ch}{rank_ch}' in '{trimmed}'")

        pieces.append({
            "piece": piece,
            "file": FILES.index(file_ch),
            "rank": int(rank_ch) - 1,
        })
    return pieces


def position_to_fen(white_pieces, black_pieces, to_move="white", castling=None):
    board = [[None] * 8 for _ in range(8)]
    occupied = set()

    def place(pieces, is_white):
        for p in pieces:
            board_rank = 7 - p["rank"]
            key = (p["file"], board_rank)
            if key in occupied:
                raise ValueError(f"Duplicate square: {FILES[p['file']]}{p['rank'] + 1}")
            occupied.add(key)
            board[board_rank][p["file"]] = p["piece"].upper() if is_white else p["piece"].lower()

    place(white_pieces, True)
    place(black_pieces, False)

    fen_ranks = []
    for r in range(8):
        rank_str = ""
        empty = 0
        for f in range(8):
            if board[r][f]:
                if empty > 0:
                    rank_str += str(empty)
                    empty = 0
                rank_str += board[r][f]
            else:
                empty += 1
        if empty > 0:
            rank_str += str(empty)
        fen_ranks.append(rank_str)

    side = "b" if to_move == "black" else "w"
    castling_str = castling or "-"
    return f"{'/'.join(fen_ranks)} {side} {castling_str} - 0 1"


# --- Commands ---

def cmd_validate(args):
    result = validate_fen(args.fen)
    print(json.dumps(result, indent=2))


def cmd_url(args):
    validation = validate_fen(args.fen)
    if not validation["valid"]:
        print(json.dumps({"error": validation["error"]}), file=sys.stderr)
        sys.exit(1)
    url = build_url(validation["fen"], args.color)
    print(json.dumps({"url": url, "fen": validation["fen"]}, indent=2))


def cmd_open(args):
    validation = validate_fen(args.fen)
    if not validation["valid"]:
        print(json.dumps({"error": validation["error"]}), file=sys.stderr)
        sys.exit(1)
    url = build_url(validation["fen"], args.color)

    system = platform.system()
    if system == "Darwin":
        open_cmd = ["open", url]
    elif system == "Windows":
        open_cmd = ["start", url]
    else:
        open_cmd = ["xdg-open", url]

    try:
        subprocess.run(open_cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        print(json.dumps({"opened": True, "url": url, "platform": system.lower()}, indent=2))
    except (subprocess.CalledProcessError, FileNotFoundError):
        print(json.dumps({
            "opened": False,
            "url": url,
            "platform": system.lower(),
            "error": f"Failed to open browser. Visit: {url}",
        }, indent=2))


def cmd_board(args):
    validation = validate_fen(args.fen)
    if not validation["valid"]:
        print(f"Error: {validation['error']}", file=sys.stderr)
        sys.exit(1)
    print(render_board(validation["fen"], args.color))


def cmd_position(args):
    if not args.white and not args.black:
        print(json.dumps({"error": "Provide at least --white or --black piece list"}), file=sys.stderr)
        sys.exit(1)

    white_pieces = parse_piece_list(args.white) if args.white else []
    black_pieces = parse_piece_list(args.black) if args.black else []
    to_move = args.to_move or "white"
    castling = args.castling if args.castling else None

    fen = position_to_fen(white_pieces, black_pieces, to_move, castling)
    url = build_url(fen)

    print(json.dumps({
        "fen": fen,
        "url": url,
        "pieces": {"white": len(white_pieces), "black": len(black_pieces)},
    }, indent=2))


# --- Main ---

def main():
    parser = argparse.ArgumentParser(description="Chess helper for Synapse")
    subparsers = parser.add_subparsers(dest="command", required=True)

    # validate command
    p_validate = subparsers.add_parser("validate", help="Validate a FEN string")
    p_validate.add_argument("--fen", required=True, help="FEN string to validate")

    # url command
    p_url = subparsers.add_parser("url", help="Generate Lichess analysis URL")
    p_url.add_argument("--fen", required=True, help="FEN string")
    p_url.add_argument("--color", choices=["white", "black"], default=None, help="Board orientation")

    # open command
    p_open = subparsers.add_parser("open", help="Open position in browser on Lichess")
    p_open.add_argument("--fen", required=True, help="FEN string")
    p_open.add_argument("--color", choices=["white", "black"], default=None, help="Board orientation")

    # board command
    p_board = subparsers.add_parser("board", help="Render Unicode board to stdout")
    p_board.add_argument("--fen", required=True, help="FEN string")
    p_board.add_argument("--color", choices=["white", "black"], default=None, help="Board orientation")

    # position command
    p_position = subparsers.add_parser("position", help="Build FEN from piece lists")
    p_position.add_argument("--white", default=None, help='White pieces, e.g. "Ke1,Qd1,Ra1"')
    p_position.add_argument("--black", default=None, help='Black pieces, e.g. "Ke8,Qd8"')
    p_position.add_argument("--to-move", dest="to_move", choices=["white", "black"], default=None, help="Side to move")
    p_position.add_argument("--castling", default=None, help="Castling rights, e.g. KQkq or -")

    args = parser.parse_args()

    if args.command == "validate":
        cmd_validate(args)
    elif args.command == "url":
        cmd_url(args)
    elif args.command == "open":
        cmd_open(args)
    elif args.command == "board":
        cmd_board(args)
    elif args.command == "position":
        cmd_position(args)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)
